plugins/events/natsjs/nats.go

```diff
@@ -0,0 +1,227 @@
+	"go-micro.dev/v4/logger"
+)
+
+const (
+	defaultClusterID      = "micro"
+	reconnectLoopDuration = 10 * time.Second
+)
+
+// NewStream returns an initialized nats stream or an error if the connection to the nats
+// server could not be established
+func NewStream(opts ...Option) (events.Stream, error) {
+	// parse the options
+	options := Options{
+		ClientID:  uuid.New().String(),
+		ClusterID: defaultClusterID,
+	}
+	for _, o := range opts {
+		o(&options)
+	}
+
+	s := &stream{opts: options}
+	natsJetStreamCtx, err := connectToNatsJetStream(options)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to nats cluster %v: %v", options.ClusterID, err)
+	}
+	s.natsJetStreamCtx = natsJetStreamCtx
+	return s, nil
+}
+
+type stream struct {
+	opts             Options
+	natsJetStreamCtx nats.JetStreamContext
+}
+
+func connectToNatsJetStream(options Options) (nats.JetStreamContext, error) {
+	nopts := nats.GetDefaultOptions()
+	if options.TLSConfig != nil {
+		nopts.Secure = true
+		nopts.TLSConfig = options.TLSConfig
+	}
+	if len(options.Address) > 0 {
+		nopts.Servers = strings.Split(options.Address, ",")
+	}
+
+	conn, err := nopts.Connect()
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to nats at %v with tls enabled (%v): %v", options.Address, nopts.TLSConfig != nil, err)
+	}
+
+	js, err := conn.JetStream()
+	if err != nil {
+		return nil, fmt.Errorf("error while obtaining JetStream context: %v", err)
+	}
+
+	return js, nil
+}
+
+// Publish a message to a topic
+func (s *stream) Publish(topic string, msg interface{}, opts ...events.PublishOption) error {
+	// validate the topic
+	if len(topic) == 0 {
+		return events.ErrMissingTopic
+	}
+
+	// parse the options
+	options := events.PublishOptions{
+		Timestamp: time.Now(),
+	}
+	for _, o := range opts {
+		o(&options)
+	}
+
+	// encode the message if it's not already encoded
+	var payload []byte
+	if p, ok := msg.([]byte); ok {
+		payload = p
+	} else {
+		p, err := json.Marshal(msg)
+		if err != nil {
+			return events.ErrEncodingMessage
+		}
+		payload = p
+	}
+
+	// construct the event
+	event := &events.Event{
+		ID:        uuid.New().String(),
+		Topic:     topic,
+		Timestamp: options.Timestamp,
+		Metadata:  options.Metadata,
+		Payload:   payload,
+	}
+
+	// serialize the event to bytes
+	bytes, err := json.Marshal(event)
+	if err != nil {
+		return errors.Wrap(err, "Error encoding event")
+	}
+
+	pubOpts := []nats.PubOpt{}
+
+	// publish the event to the topic's channel
+	if _, err := s.natsJetStreamCtx.PublishAsync(event.Topic, bytes, pubOpts...); err != nil {
+		return errors.Wrap(err, "Error publishing message to topic")
+	}
+
+	return nil
+}
+
+// Consume to a topic
+func (s *stream) Consume(topic string, opts ...events.ConsumeOption) (<-chan events.Event, error) {
+	// validate the topic
+	if len(topic) == 0 {
+		return nil, events.ErrMissingTopic
+	}
+
+	// parse the options
+	options := events.ConsumeOptions{
+		Group: uuid.New().String(),
```

Comment:
Group name should be defineable. I want multiple services listening in the same group

Reply:
That's just a "default" group if you don't provide one via the opts.

If you provide one, it'll be set in the next few lines:
```
	for _, o := range opts {
		o(&options)
	}
```